 # Problem 1

## a)

## Take this problem as a max flow problem with multiple sinks and draw the network and label the capacities of the arcs:

<img src="Homework3Q1.bmp">

## b)

In [1]:
using JuMP, Clp, NamedArrays
trucks = [:A, :B, :C, :D]
loads = [:E, :F, :G, :H]
arcs = [:EA, :EC, :FB, :GA, :GB, :HA, :HC, :HD]

A = NamedArray( [1 0 0 1 0 1 0 0;0 0 1 0 1 0 0 0;0 1 0 0 0 0 1 0;0 0 0 0 0 0 0 1],(trucks, arcs), ("Trucks","Arcs"))

b = NamedArray( [7;15;9;13],(trucks), ("Trucks","Amount"))

C = NamedArray( [1 1 0 0 0 0 0 0;0 0 1 0 0 0 0 0;0 0 0 1 1 0 0 0;0 0 0 0 0 1 1 1],(loads, arcs), ("Loads","Arcs"))

d = NamedArray( [7,6,16,6],(loads), ("Loads","Amount"))

m = Model()

@variable(m, x[arcs] >= 0)

@constraint(m, truckconst[i in trucks], sum(A[i,j]*x[j] for j in arcs) <= b[i])

@constraint(m, loadconst[i in loads], sum(C[i,j]*x[j] for j in arcs) >= d[i])

@objective(m, Max, sum(A[i,j]*x[j] for i in trucks, j in arcs))

set_optimizer(m, Clp.Optimizer)

optimize!(m)

println("Maximum load:", objective_value(m))
println("Trucks used:")
println("A:", value.(x[:EA]) + value.(x[:GA]) + value.(x[:HA]))
println("B:", value.(x[:FB]) + value.(x[:GB]))
println("C:", value.(x[:EC]) + value.(x[:HC]))
println("D:", value.(x[:HD]))

println("Loads used:")
println("E:", value.(x[:EA]) + value.(x[:EC]))
println("F:", value.(x[:FB]))
println("G:", value.(x[:GA]) + value.(x[:GB]))
println("H:", value.(x[:HA]) + value.(x[:HC]) + value.(x[:HD]))

println("Loads on arcs:")
println("EA:", value.(x[:EA]))
println("EC:", value.(x[:EC]))
println("FB:", value.(x[:FB]))
println("GA:", value.(x[:GA]))
println("GB:", value.(x[:GB]))
println("HA:", value.(x[:HA]))
println("HC:", value.(x[:HC]))
println("HD:", value.(x[:HD]))

Maximum load:44.0
Trucks used:
A:7.0
B:15.0
C:9.0
D:13.0
Loads used:
E:9.0
F:6.0
G:16.0
H:13.0
Loads on arcs:
EA:0.0
EC:9.0
FB:6.0
GA:7.0
GB:9.0
HA:0.0
HC:0.0
HD:13.0
Coin0506I Presolve 0 (-8) rows, 0 (-8) columns and 0 (-16) elements
Clp3002W Empty problem - 0 rows, 0 columns and 0 elements
Clp0000I Optimal - objective value 44
Coin0511I After Postsolve, objective 44, infeasibilities - dual 0 (0), primal 0 (0)
Clp0032I Optimal objective 44 - 0 iterations time 0.002, Presolve 0.00


## c)

In [2]:
using JuMP, Clp

a = [1 0 0 1 0 1 0 0;0 0 1 0 1 0 0 0;0 1 0 0 0 0 1 0;0 0 0 0 0 0 0 1]

m = Model()

@variable(m, x[1:8] >= 0)

@constraint(m, A, x[1] + x[4] + x[6] <= 7)

@constraint(m, B, x[3] + x[5] <= 15)

@constraint(m, C, x[2] + x[7] <= 9)

@constraint(m, D, x[8] <= 13)

@constraint(m, E, x[1] + x[2] >= 7)

@constraint(m, F, x[3] >= 6)

@constraint(m, G, x[4] + x[5] >= 16)

@constraint(m, H, x[6] + x[7] + x[8] >= 6)

@objective(m, Max, sum(a[i,j]*x[j] for i in 1:4, j in 1:8))

set_optimizer(m, Clp.Optimizer)

optimize!(m)

println("dual for A:", dual(A))
println("dual for B:", dual(B))
println("dual for C:", dual(C))
println("dual for D:", dual(D))
println("dual for E:", dual(E))
println("dual for F:", dual(F))
println("dual for G:", dual(G))
println("dual for H:", dual(H))

dual for A:-1.0
dual for B:-1.0
dual for C:-1.0
dual for D:-1.0
dual for E:0.0
dual for F:0.0
dual for G:0.0
dual for H:0.0
Coin0506I Presolve 0 (-8) rows, 0 (-8) columns and 0 (-16) elements
Clp3002W Empty problem - 0 rows, 0 columns and 0 elements
Clp0000I Optimal - objective value 44
Coin0511I After Postsolve, objective 44, infeasibilities - dual 0 (0), primal 0 (0)
Clp0032I Optimal objective 44 - 0 iterations time 0.002, Presolve 0.00


## The minimum cost will be 4.

 # Problem 2

## a)

## this problem can be illustrated as:$$ max\ p=c^Tx$$
## $$ s.t.Ax\leq b $$
## $$ Fx=g $$
## $$ x\geq0 $$
## where:$x=[x_1;x_2;x_3], c=[2;1;4], A=[1,2,2], F=[2,-1,3], b=[3], g=[3]$

## Then the dual problem is:$$ min\ d=b^T\lambda+g^T\mu$$
## $$ s.t. \lambda\geq0 $$
## $$ \mu\ free $$
## $$ A^T\lambda+F^T\mu\geq c$$
## because $\lambda=[\lambda_1], \mu=[\mu_1]$, this can be rewritten as:
## $$ min\ d=3\lambda_1+3\mu_1$$
## $$ s.t. \begin{cases} \lambda_1\geq0 \\
\mu_1\ free \\
1\lambda_1+2\mu_1\geq 2 \\
2\lambda_1-1\mu_1\geq 1 \\
2\lambda_1+3\mu_1\geq 4\end{cases}$$

## Since we know for optimal solution the relation the strong condition $p^*=d^*$ holds. For the dual problem, $\lambda_1=1,\mu_1=1$is a feasilble solution. Then we have $p^*\leq6$. In fact:

In [3]:
using JuMP, Clp

m = Model()

@variable(m, x >= 0)
@variable(m, y)

@objective(m, Min, 3*x+3*y)

@constraint(m, x + 2*y >= 2)

@constraint(m, 2*x - 1*y >= 1)

@constraint(m, 2*x + 3*y >= 4)

set_optimizer(m, Clp.Optimizer)

optimize!(m)

println("lambda_1:", value(x))
println("mu_1:", value(y))
println("Minimum", objective_value(m))

lambda_1:0.8749999999999999
mu_1:0.75
Minimum4.875
Coin0506I Presolve 3 (0) rows, 2 (0) columns and 6 (0) elements
Clp0006I 0  Obj 0 Primal inf 2.833333 (3) Dual inf 0.0299999 (1) w.o. free dual inf (0)
Clp0006I 2  Obj 4.875
Clp0000I Optimal - objective value 4.875
Clp0032I Optimal objective 4.875 - 2 iterations time 0.002


## b)

## set the decision variables for dual problem as $[y_1,y_2,y_3,y_4]$
## $$ \min_{y_1,y_2,y_3}\ d=4y_1+36y_2-16y_3$$
## $$ s.t.{y_1\leq6\\
-y_1-4y_2=-2\\
6y_2-y_3\geq4\\
-y_3\geq1}$$
## $$ y_1 free, y_2\geq0, y_3\leq0 $$

## with $x=(0,-4,\frac{10}{3},\frac{38}{3})$, it is easy to find that the $x_3>0$ and $x_4>0$ are satisfied. This means the $6y_2-y_3\geq4, -y_3\geq1$ are just equals. So we have:
## $$ \left[ \begin{array} {ccc}-1&-4&0\\0&6&-1\\0&0&-1\end{array} \right]\left[ \begin{array} {c}y_1\\y_2\\y_3\end{array} \right]=\left[ \begin{array} {c}-2\\4\\1\end{array} \right]$$
## So $$\left[ \begin{array} {c}y_1\\y_2\\y_3\end{array} \right]=\left[ \begin{array} {c}0\\0.5\\-1\end{array} \right]$$
## Since the constrains of dual problem is satsified, $x=(0,-4,\frac{10}{3},\frac{38}{3})$ is a optimal solution.

## with $y=(0,\frac{1}{2},-1)$, it is easy to find that the $y_2>0$ and $y_3<0$ are satisfied. This means the $-4x_2+6x_3\leq36, -x_3-x_4\geq-16$ are just equals. So we have:
## $$ \left[ \begin{array} {cccc}1&-1&0&0\\0&-4&6&0\\0&0&-1&-1\end{array} \right]\left[ \begin{array} {c}x_1\\x_2\\x_3\\x_4\end{array} \right]=\left[ \begin{array} {c}4\\36\\-16\end{array} \right]$$
## Also, we can find slackness in $y_1\leq6$, which means $x_1=0$. And we have:
## $$\left[ \begin{array} {c}x_1\\x_2\\x_3\\x_4\end{array} \right]=\left[ \begin{array} {c}0\\-4\\\frac{10}{3}\\\frac{38}{3}\end{array} \right]$$
## In all, $x=(0,-4,\frac{10}{3},\frac{38}{3})$ is a optimal solution.

 # Problem 3

## a)

In [4]:
using JuMP, Clp

m = Model()

@variable(m, x1 >= 0)
@variable(m, x2 >= 0)
@variable(m, x3 >= 0)
@variable(m, x4 >= 0)

@objective(m, Max, 60*x1 + 120*x2 + 200*x3 + 300*x4)

@constraint(m, 2*x1 + 3*x2 + 3*x3 + 5*x4 <= 12000)

@constraint(m, 5*x1 + 5*x2 + 10*x3 + 15*x4 <= 32000)

@constraint(m, 0.25*x1 + 1*x2 + 2*x3 + 3.5*x4 <= 5000)

set_optimizer(m, Clp.Optimizer)

optimize!(m)

println("pandas:", value(x1))
println("panthers:", value(x2))
println("pangolins:", value(x3))
println("penguins:", value(x4))
println("Maximum", objective_value(m))

pandas:1866.6666666666665
panthers:977.777777777778
pangolins:1777.7777777777774
penguins:0.0
Maximum584888.8888888889
Coin0506I Presolve 3 (0) rows, 4 (0) columns and 12 (0) elements
Clp0006I 0  Obj 0 Dual inf 380.83786 (4)
Clp0006I 3  Obj 584888.89
Clp0000I Optimal - objective value 584888.89
Clp0032I Optimal objective 584888.8889 - 3 iterations time 0.002


## set the decision variables for dual problem as $y=[y_1,y_2,y_3]$
## $$ \min_{y_1,y_2,y_3}\ d=12000y_1+32000y_2+5000y_3$$
## $$ s.t.{2y_1+5y_2+0.25y_3\geq60\\
3y_1+5y_2+1y_3\geq120\\
3y_1+10y_2+2y_3\geq200\\
5y_1+15y_2+3.5y_3\geq300\\}$$
$$ y_1\geq0, y_2\geq0, y_3\geq0 $$

In [5]:
using JuMP, Clp

m = Model()

@variable(m, y1 >= 0)
@variable(m, y2 >= 0)
@variable(m, y3 >= 0)

@objective(m, Min, 12000*y1 + 32000*y2 + 5000*y3)

@constraint(m, 2*y1 + 5*y2 + 0.25*y3 >= 60)

@constraint(m, 3*y1 + 5*y2 + 1*y3 >= 120)

@constraint(m, 3*y1 + 10*y2 + 2*y3 >= 200)

@constraint(m, 5*y1 + 15*y2 + 3.5*y3 >= 300)

set_optimizer(m, Clp.Optimizer)

optimize!(m)

println("y1=", value(y1))
println("y2=", value(y2))
println("y3=", value(y3))
println("Minimum", objective_value(m))

y1=13.333333333333332
y2=3.555555555555556
y3=62.22222222222221
Minimum584888.8888888888
Coin0506I Presolve 4 (0) rows, 3 (0) columns and 12 (0) elements
Clp0006I 0  Obj 0 Primal inf 76 (4)
Clp0006I 3  Obj 584888.89
Clp0000I Optimal - objective value 584888.89
Clp0032I Optimal objective 584888.8889 - 3 iterations time 0.002


## b)

In [6]:
x1=1866.6666666666665
x2=977.777777777778
x3=1777.7777777777774
x4=0.0
println("Labor used in hour:",2*x1 + 3*x2 + 3*x3 + 5*x4)

Labor used in hour:12000.0


## Here the labor hours are fully used to reach the optimal solution. Also, we have the optimal solution for dual problem with $y1=13.333333333333332$. so the maximum amount Assemble-an-Animal should be willing to pay for an additional hour of labor is 13.3333\$.

## c)

## The profit per panther plushie is increased by $10$ and pangolin plushie is increased by $15$

In [7]:
using JuMP, Clp

m = Model()

@variable(m, x1 >= 0)
@variable(m, x2 >= 0)
@variable(m, x3 >= 0)
@variable(m, x4 >= 0)

@objective(m, Max, 60*x1 + 130*x2 + 215*x3 + 300*x4)

@constraint(m, 2*x1 + 3*x2 + 3*x3 + 5*x4 <= 12000)

@constraint(m, 5*x1 + 5*x2 + 10*x3 + 15*x4 <= 32000)

@constraint(m, 0.25*x1 + 1*x2 + 2*x3 + 3.5*x4 <= 5000)

set_optimizer(m, Clp.Optimizer)

optimize!(m)

println("pandas:", value(x1))
println("panthers:", value(x2))
println("pangolins:", value(x3))
println("penguins:", value(x4))
println("Maximum", objective_value(m))

pandas:1866.6666666666665
panthers:977.777777777778
pangolins:1777.7777777777774
penguins:0.0
Maximum621333.3333333333
Coin0506I Presolve 3 (0) rows, 4 (0) columns and 12 (0) elements
Clp0006I 0  Obj 0 Dual inf 397.20761 (4)
Clp0006I 3  Obj 621333.33
Clp0000I Optimal - objective value 621333.33
Clp0032I Optimal objective 621333.3333 - 3 iterations time 0.002


## old cases without increase in profit:
## pandas:1866.6666666666665
## panthers:977.777777777778
## pangolins:1777.7777777777774
## penguins:0.0

In [8]:
using JuMP, Clp

m = Model()

@variable(m, y1 >= 0)
@variable(m, y2 >= 0)
@variable(m, y3 >= 0)

@objective(m, Min, 12000*y1 + 32000*y2 + 5000*y3)

@constraint(m, 2*y1 + 5*y2 + 0.25*y3 >= 60)

@constraint(m, 3*y1 + 5*y2 + 1*y3 >= 130)

@constraint(m, 3*y1 + 10*y2 + 2*y3 >= 215)

@constraint(m, 5*y1 + 15*y2 + 3.5*y3 >= 300)

set_optimizer(m, Clp.Optimizer)

optimize!(m)

println("y1=", value(y1))
println("y2=", value(y2))
println("y3=", value(y3))
println("Minimum", objective_value(m))

y1=14.999999999999998
y2=2.333333333333334
y3=73.33333333333331
Minimum621333.3333333333
Coin0506I Presolve 4 (0) rows, 3 (0) columns and 12 (0) elements
Clp0006I 0  Obj 0 Primal inf 79.5 (4)
Clp0006I 3  Obj 621333.33
Clp0000I Optimal - objective value 621333.33
Clp0032I Optimal objective 621333.3333 - 3 iterations time 0.002


## Summarizing the above calculation, we have the following form of $z^*_{NEW}$:
## $$z^*_{NEW}\geq584888.89$$

## d)

## The labor availability is decreased by 1000 hours, but the raw material availability is increased by 4000 square feet

In [9]:
using JuMP, Clp

m = Model()

@variable(m, x1 >= 0)
@variable(m, x2 >= 0)
@variable(m, x3 >= 0)
@variable(m, x4 >= 0)

@objective(m, Max, 60*x1 + 120*x2 + 200*x3 + 300*x4)

@constraint(m, 2*x1 + 3*x2 + 3*x3 + 5*x4 <= 11000)

@constraint(m, 5*x1 + 5*x2 + 10*x3 + 15*x4 <= 36000)

@constraint(m, 0.25*x1 + 1*x2 + 2*x3 + 3.5*x4 <= 5000)

set_optimizer(m, Clp.Optimizer)

optimize!(m)

println("pandas:", value(x1))
println("panthers:", value(x2))
println("pangolins:", value(x3))
println("penguins:", value(x4))
println("Maximum", objective_value(m))

pandas:2153.8461538461543
panthers:0.0
pangolins:2230.769230769231
penguins:0.0
Maximum575384.6153846155
Coin0506I Presolve 3 (0) rows, 4 (0) columns and 12 (0) elements
Clp0006I 0  Obj 0 Dual inf 380.83786 (4)
Clp0006I 2  Obj 575384.62
Clp0000I Optimal - objective value 575384.62
Clp0032I Optimal objective 575384.6154 - 2 iterations time 0.002


In [10]:
using JuMP, Clp

m = Model()

@variable(m, y1 >= 0)
@variable(m, y2 >= 0)
@variable(m, y3 >= 0)

@objective(m, Min, 11000*y1 + 36000*y2 + 5000*y3)

@constraint(m, 2*y1 + 5*y2 + 0.25*y3 >= 60)

@constraint(m, 3*y1 + 5*y2 + 1*y3 >= 120)

@constraint(m, 3*y1 + 10*y2 + 2*y3 >= 200)

@constraint(m, 5*y1 + 15*y2 + 3.5*y3 >= 300)

set_optimizer(m, Clp.Optimizer)

optimize!(m)

println("y1=", value(y1))
println("y2=", value(y2))
println("y3=", value(y3))
println("Minimum", objective_value(m))

y1=21.538461538461537
y2=0.0
y3=67.69230769230768
Minimum575384.6153846153
Coin0506I Presolve 4 (0) rows, 3 (0) columns and 12 (0) elements
Clp0006I 0  Obj 0 Primal inf 76 (4)
Clp0006I 3  Obj 575384.62
Clp0000I Optimal - objective value 575384.62
Clp0032I Optimal objective 575384.6154 - 3 iterations time 0.002


## Summarizing the above calculation, we have the following form of $z^*_{NEW}$:
## $$z^*_{NEW}\leq584888.89$$